File: backend/app/modules/capital_ownership/progress.py
"""Section completion for Capital & Ownership — ports `frontend/lib/capital-ownership/progress.ts`.

Mirrors the IPO Setup progress model: each section resolves to
`not_started | in_progress | complete`, and an unanswered ternary is never treated as "no".
"""

from __future__ import annotations

from typing import Any

from app.modules.capital_ownership import decimal_math as dm
from app.modules.capital_ownership.constants import SECTION_IDS, SECTION_LABELS


def _filled(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return bool(value.strip())
    if isinstance(value, bool):
        return value
    if isinstance(value, list):
        return len(value) > 0
    return True


def _status_from(answered: int, total: int, extra_complete: bool = True) -> str:
    if answered == 0:
        return "not_started"
    if answered < total or not extra_complete:
        return "in_progress"
    return "complete"


def evaluate_current_capital_structure_status(payload: dict[str, Any]) -> str:
    section = payload.get("currentCapitalStructure") or {}
    equity_classes = section.get("equityClasses") or []
    equity_rows_entered = [
        item
        for item in equity_classes
        if _filled(item.get("className")) or dm.is_filled(item.get("issuedShares"))
    ]

    core = [
        _filled(section.get("asOnDate")),
        len(equity_rows_entered) > 0,
        dm.is_filled(section.get("authorisedEquityShareCapital")),
        dm.is_filled(section.get("issuedEquityShareCapital")),
        dm.is_filled(section.get("paidUpEquityShareCapital")),
        _filled(section.get("allSharesFullyPaidUp")),
        _filled(section.get("shareCapitalMatchesMcaRecords")),
        _filled(section.get("dematStatusOverall")),
    ]
    answered = sum(1 for value in core if value)

    classes_complete = all(
        dm.is_filled(item.get("faceValuePerShare"))
        and dm.is_filled(item.get("issuedShares"))
        and dm.is_filled(item.get("paidUpShares"))
        for item in equity_rows_entered
    )
    discrepancy_explained = section.get("shareCapitalMatchesMcaRecords") != "no" or _filled(
        section.get("discrepancyWithMcaRecordsExplanation")
    )
    partly_paid_explained = section.get("partlyPaidSharesOutstanding") != "yes" or _filled(
        section.get("partlyPaidSharesDetails")
    )

    return _status_from(
        answered,
        len(core),
        classes_complete and discrepancy_explained and partly_paid_explained,
    )


def evaluate_share_capital_history_status(payload: dict[str, Any]) -> str:
    section = payload.get("shareCapitalHistory") or {}
    core = [
        _filled(section.get("historyCoversPeriodSinceIncorporation")),
        len(section.get("capitalEvents") or []) > 0,
        _filled(section.get("allHistoricalAllotmentsDocumented")),
        _filled(section.get("historyReconciledWithMcaFilings")),
        _filled(section.get("bonusIssueInLastTwelveMonths")),
        _filled(section.get("sharesIssuedForConsiderationOtherThanCashInLastTwelveMonths")),
        _filled(section.get("anyPendingAllotments")),
    ]
    answered = sum(1 for value in core if value)

    events_complete = all(
        _filled(event.get("eventDate"))
        and _filled(event.get("eventType"))
        and (
            dm.is_filled(event.get("numberOfShares"))
            or event.get("eventType") == "increase-in-authorised-capital"
        )
        for event in (section.get("capitalEvents") or [])
    )
    gaps_explained = section.get("allHistoricalAllotmentsDocumented") != "no" or _filled(
        section.get("gapsInHistoryExplanation")
    )
    pending_explained = section.get("anyPendingAllotments") != "yes" or _filled(
        section.get("pendingAllotmentDetails")
    )

    return _status_from(answered, len(core), events_complete and gaps_explained and pending_explained)


def evaluate_shareholders_status(payload: dict[str, Any]) -> str:
    section = payload.get("shareholdersAndBeneficialOwnership") or {}
    core = [
        _filled(section.get("shareholdingAsOnDate")),
        len(section.get("shareholders") or []) > 0,
        dm.is_filled(section.get("totalNumberOfShareholders")),
        _filled(section.get("registerOfMembersMaintained")),
        _filled(section.get("significantBeneficialOwnerDeterminationCompleted")),
        _filled(section.get("foreignShareholdingExists")),
        _filled(section.get("anyShareholderAgreementsWithInvestors")),
    ]
    answered = sum(1 for value in core if value)

    shareholders_complete = all(
        _filled(item.get("name"))
        and _filled(item.get("category"))
        and dm.is_filled(item.get("equitySharesHeld"))
        for item in (section.get("shareholders") or [])
    )
    sbo_complete = section.get(
        "significantBeneficialOwnerDeterminationCompleted"
    ) != "yes" or len(section.get("beneficialOwners") or []) > 0
    foreign_complete = section.get("foreignShareholdingExists") != "yes" or _filled(
        section.get("foreignDirectInvestmentComplianceConfirmed")
    )

    return _status_from(
        answered, len(core), shareholders_complete and sbo_complete and foreign_complete
    )


def evaluate_promoters_and_control_status(payload: dict[str, Any]) -> str:
    section = payload.get("promotersAndControl") or {}
    has_promoter = section.get("companyHasIdentifiedPromoter") == "yes"
    core = [
        _filled(section.get("companyHasIdentifiedPromoter")),
        len(section.get("promoters") or []) > 0
        if has_promoter
        else _filled(section.get("noPromoterExplanation")),
        _filled(section.get("promoterIdentificationComplete")),
        _filled(section.get("promoterGroupIdentificationComplete")),
        _filled(section.get("anyPersonExercisingControlWithoutShareholding")),
        _filled(section.get("changeInControlInLastThreeYears")),
    ]
    answered = sum(1 for value in core if value)

    promoters_complete = all(
        _filled(item.get("name"))
        and _filled(item.get("promoterType"))
        and _filled(item.get("basisOfPromoterStatus"))
        for item in (section.get("promoters") or [])
    )
    group_complete = all(
        _filled(item.get("name")) and _filled(item.get("relationshipToPromoter"))
        for item in (section.get("promoterGroupMembers") or [])
    )
    arrangements_complete = all(
        _filled(item.get("arrangementType")) and _filled(item.get("partiesInvolved"))
        for item in (section.get("controlArrangements") or [])
    )
    control_explained = section.get(
        "anyPersonExercisingControlWithoutShareholding"
    ) != "yes" or _filled(section.get("controlWithoutShareholdingDetails"))
    change_explained = section.get("changeInControlInLastThreeYears") != "yes" or _filled(
        section.get("changeInControlDetails")
    )

    return _status_from(
        answered,
        len(core),
        promoters_complete
        and group_complete
        and arrangements_complete
        and control_explained
        and change_explained,
    )


def evaluate_pre_post_issue_status(payload: dict[str, Any]) -> str:
    section = payload.get("preAndPostIssueOwnership") or {}
    has_overlays = len(section.get("shareholderOverlays") or []) > 0
    core = [
        has_overlays,
        _filled(section.get("preIssueCapitalConfirmedWithLeadManager")),
        _filled(section.get("sellingShareholderConsentsObtained")),
        _filled(section.get("anyExpectedPreIssueTransfers")),
    ]
    answered = sum(1 for value in core if value)

    overlays_complete = all(
        _filled(overlay.get("shareholderId"))
        and (
            dm.is_filled(overlay.get("sharesOfferedForSale"))
            or dm.is_filled(overlay.get("otherExpectedPreIssueTransfer"))
        )
        for overlay in (section.get("shareholderOverlays") or [])
    )
    sellers_complete = section.get("sellingShareholderConsentsObtained") != "yes" or _filled(
        section.get("sellingShareholderEligibilityConfirmed")
    )
    transfers_explained = section.get("anyExpectedPreIssueTransfers") != "yes" or _filled(
        section.get("expectedPreIssueTransferDetails")
    )

    return _status_from(
        answered, len(core), overlays_complete and sellers_complete and transfers_explained
    )


def evaluate_promoter_contribution_status(payload: dict[str, Any]) -> str:
    section = payload.get("promoterContributionLockInAndEncumbrances") or {}
    applicable = section.get("minimumPromoterContributionApplicable") != "no"
    core = [
        _filled(section.get("minimumPromoterContributionApplicable")),
        len(section.get("contributionLots") or []) > 0
        if applicable
        else _filled(section.get("exemptionFromMinimumContributionClaimed")),
        _filled(section.get("contributionBroughtInBeforeIssueOpening")),
        _filled(section.get("anyEncumbranceOnPromoterShares")),
        _filled(section.get("entirePreIssueCapitalLockInUnderstood")),
        _filled(section.get("sharesIneligibleForContributionExist")),
    ]
    answered = sum(1 for value in core if value)

    lots_complete = all(
        dm.is_filled(lot.get("numberOfShares"))
        and _filled(lot.get("dateOfAcquisition"))
        and _filled(lot.get("modeOfAcquisition"))
        and _filled(lot.get("eligibleForMinimumPromoterContribution"))
        for lot in (section.get("contributionLots") or [])
    )
    encumbrances_complete = all(
        _filled(item.get("encumbranceType")) and dm.is_filled(item.get("numberOfSharesEncumbered"))
        for item in (section.get("encumbrances") or [])
    )
    encumbrance_answered = section.get("anyEncumbranceOnPromoterShares") != "yes" or len(
        section.get("encumbrances") or []
    ) > 0
    ineligible_explained = section.get("sharesIneligibleForContributionExist") != "yes" or _filled(
        section.get("ineligibleSharesDetails")
    )
    exemption_explained = section.get(
        "exemptionFromMinimumContributionClaimed"
    ) != "yes" or _filled(section.get("exemptionBasis"))

    return _status_from(
        answered,
        len(core),
        lots_complete
        and encumbrances_complete
        and encumbrance_answered
        and ineligible_explained
        and exemption_explained,
    )


def evaluate_outstanding_securities_status(payload: dict[str, Any]) -> str:
    section = payload.get("outstandingSecuritiesTransactionsAndConfirmations") or {}
    confirmations = list((section.get("confirmations") or {}).values())
    confirmations_checked = sum(1 for value in confirmations if value)

    core = [
        _filled(section.get("anyOutstandingConvertibleInstruments")),
        _filled(section.get("anyTransactionsInLastEighteenMonths")),
        _filled(section.get("allSharesDematerialisedBeforeFiling")),
        _filled(section.get("anyPendingShareTransfers")),
        _filled(section.get("anyDisputesOverTitleToShares")),
        confirmations_checked > 0,
    ]
    answered = sum(1 for value in core if value)

    instruments = section.get("outstandingInstruments") or []
    instruments_complete = section.get("anyOutstandingConvertibleInstruments") != "yes" or (
        len(instruments) > 0
        and all(
            _filled(item.get("instrumentType"))
            and dm.is_filled(item.get("potentialEquitySharesOnConversion"))
            for item in instruments
        )
    )
    transactions = section.get("recentTransactions") or []
    transactions_complete = section.get("anyTransactionsInLastEighteenMonths") != "yes" or (
        len(transactions) > 0
        and all(
            _filled(item.get("transactionDate"))
            and _filled(item.get("transactionType"))
            and dm.is_filled(item.get("numberOfShares"))
            for item in transactions
        )
    )
    pending_explained = section.get("anyPendingShareTransfers") != "yes" or _filled(
        section.get("pendingShareTransferDetails")
    )
    disputes_explained = section.get("anyDisputesOverTitleToShares") != "yes" or _filled(
        section.get("titleDisputeDetails")
    )
    confirmations_complete = confirmations_checked == len(confirmations)

    return _status_from(
        answered,
        len(core),
        instruments_complete
        and transactions_complete
        and pending_explained
        and disputes_explained
        and confirmations_complete,
    )


_EVALUATORS = {
    "current-capital-structure": evaluate_current_capital_structure_status,
    "share-capital-history": evaluate_share_capital_history_status,
    "shareholders-beneficial-ownership": evaluate_shareholders_status,
    "promoters-and-control": evaluate_promoters_and_control_status,
    "pre-post-issue-ownership": evaluate_pre_post_issue_status,
    "promoter-contribution-lock-in": evaluate_promoter_contribution_status,
    "outstanding-securities-confirmations": evaluate_outstanding_securities_status,
}


def calculate_progress(payload: dict[str, Any]) -> dict[str, Any]:
    sections = {section_id: _EVALUATORS[section_id](payload) for section_id in SECTION_IDS}
    statuses = list(sections.values())
    sections_complete = sum(1 for status in statuses if status == "complete")
    total_sections = len(statuses)
    if sections_complete == total_sections:
        overall = "complete"
    elif any(status != "not_started" for status in statuses):
        overall = "in_progress"
    else:
        overall = "not_started"
    return {
        "sections": sections,
        "sectionsComplete": sections_complete,
        "totalSections": total_sections,
        "overallStatus": overall,
    }


def list_missing_required(payload: dict[str, Any]) -> list[str]:
    progress = calculate_progress(payload)
    missing: list[str] = []
    for section_id, status in progress["sections"].items():
        if status != "complete":
            missing.append(f"{SECTION_LABELS[section_id]} incomplete")
    return missing

File: frontend/lib/capital-ownership/progress.ts
/**
 * Section completion for Capital & Ownership.
 *
 * Mirrors the IPO Setup progress model: each section resolves to
 * `not_started | in_progress | complete`, and an unanswered ternary is never treated as "no".
 */

import { isFilledDecimal } from '@/lib/capital-ownership/decimal';
import { CAPITAL_OWNERSHIP_SECTION_LABELS } from '@/lib/capital-ownership/options';
import type {
  CapitalOwnershipPayload,
  CapitalOwnershipProgress,
  CapitalOwnershipSectionId,
  SectionStatus,
} from '@/lib/capital-ownership/types';

function filled(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === 'string') return value.trim().length > 0;
  if (typeof value === 'boolean') return value;
  if (Array.isArray(value)) return value.length > 0;
  return true;
}

function statusFrom(answered: number, total: number, extraComplete = true): SectionStatus {
  if (answered === 0) return 'not_started';
  if (answered < total || !extraComplete) return 'in_progress';
  return 'complete';
}

export function evaluateCurrentCapitalStructureStatus(
  payload: CapitalOwnershipPayload,
): SectionStatus {
  const section = payload.currentCapitalStructure;
  const equityRowsEntered = section.equityClasses.filter(
    (item) => filled(item.className) || isFilledDecimal(item.issuedShares),
  );

  const core = [
    filled(section.asOnDate),
    equityRowsEntered.length > 0,
    isFilledDecimal(section.authorisedEquityShareCapital),
    isFilledDecimal(section.issuedEquityShareCapital),
    isFilledDecimal(section.paidUpEquityShareCapital),
    filled(section.allSharesFullyPaidUp),
    filled(section.shareCapitalMatchesMcaRecords),
    filled(section.dematStatusOverall),
  ];
  const answered = core.filter(Boolean).length;

  const classesComplete = equityRowsEntered.every(
    (item) =>
      isFilledDecimal(item.faceValuePerShare) &&
      isFilledDecimal(item.issuedShares) &&
      isFilledDecimal(item.paidUpShares),
  );
  const discrepancyExplained =
    section.shareCapitalMatchesMcaRecords !== 'no' ||
    filled(section.discrepancyWithMcaRecordsExplanation);
  const partlyPaidExplained =
    section.partlyPaidSharesOutstanding !== 'yes' || filled(section.partlyPaidSharesDetails);

  return statusFrom(
    answered,
    core.length,
    classesComplete && discrepancyExplained && partlyPaidExplained,
  );
}

export function evaluateShareCapitalHistoryStatus(payload: CapitalOwnershipPayload): SectionStatus {
  const section = payload.shareCapitalHistory;
  const core = [
    filled(section.historyCoversPeriodSinceIncorporation),
    section.capitalEvents.length > 0,
    filled(section.allHistoricalAllotmentsDocumented),
    filled(section.historyReconciledWithMcaFilings),
    filled(section.bonusIssueInLastTwelveMonths),
    filled(section.sharesIssuedForConsiderationOtherThanCashInLastTwelveMonths),
    filled(section.anyPendingAllotments),
  ];
  const answered = core.filter(Boolean).length;

  const eventsComplete = section.capitalEvents.every(
    (event) =>
      filled(event.eventDate) &&
      filled(event.eventType) &&
      (isFilledDecimal(event.numberOfShares) || event.eventType === 'increase-in-authorised-capital'),
  );
  const gapsExplained =
    section.allHistoricalAllotmentsDocumented !== 'no' || filled(section.gapsInHistoryExplanation);
  const pendingExplained =
    section.anyPendingAllotments !== 'yes' || filled(section.pendingAllotmentDetails);

  return statusFrom(answered, core.length, eventsComplete && gapsExplained && pendingExplained);
}

export function evaluateShareholdersStatus(payload: CapitalOwnershipPayload): SectionStatus {
  const section = payload.shareholdersAndBeneficialOwnership;
  const core = [
    filled(section.shareholdingAsOnDate),
    section.shareholders.length > 0,
    isFilledDecimal(section.totalNumberOfShareholders),
    filled(section.registerOfMembersMaintained),
    filled(section.significantBeneficialOwnerDeterminationCompleted),
    filled(section.foreignShareholdingExists),
    filled(section.anyShareholderAgreementsWithInvestors),
  ];
  const answered = core.filter(Boolean).length;

  const shareholdersComplete = section.shareholders.every(
    (item) => filled(item.name) && filled(item.category) && isFilledDecimal(item.equitySharesHeld),
  );
  const sboComplete =
    section.significantBeneficialOwnerDeterminationCompleted !== 'yes' ||
    section.beneficialOwners.length > 0;
  const foreignComplete =
    section.foreignShareholdingExists !== 'yes' ||
    filled(section.foreignDirectInvestmentComplianceConfirmed);

  return statusFrom(answered, core.length, shareholdersComplete && sboComplete && foreignComplete);
}

export function evaluatePromotersAndControlStatus(payload: CapitalOwnershipPayload): SectionStatus {
  const section = payload.promotersAndControl;
  const hasPromoter = section.companyHasIdentifiedPromoter === 'yes';
  const core = [
    filled(section.companyHasIdentifiedPromoter),
    hasPromoter ? section.promoters.length > 0 : filled(section.noPromoterExplanation),
    filled(section.promoterIdentificationComplete),
    filled(section.promoterGroupIdentificationComplete),
    filled(section.anyPersonExercisingControlWithoutShareholding),
    filled(section.changeInControlInLastThreeYears),
  ];
  const answered = core.filter(Boolean).length;

  const promotersComplete = section.promoters.every(
    (item) => filled(item.name) && filled(item.promoterType) && filled(item.basisOfPromoterStatus),
  );
  const groupComplete = section.promoterGroupMembers.every(
    (item) => filled(item.name) && filled(item.relationshipToPromoter),
  );
  const arrangementsComplete = section.controlArrangements.every(
    (item) => filled(item.arrangementType) && filled(item.partiesInvolved),
  );
  const controlExplained =
    section.anyPersonExercisingControlWithoutShareholding !== 'yes' ||
    filled(section.controlWithoutShareholdingDetails);
  const changeExplained =
    section.changeInControlInLastThreeYears !== 'yes' || filled(section.changeInControlDetails);

  return statusFrom(
    answered,
    core.length,
    promotersComplete && groupComplete && arrangementsComplete && controlExplained && changeExplained,
  );
}

export function evaluatePrePostIssueStatus(payload: CapitalOwnershipPayload): SectionStatus {
  const section = payload.preAndPostIssueOwnership;
  const hasOverlays = section.shareholderOverlays.length > 0;
  const core = [
    hasOverlays,
    filled(section.preIssueCapitalConfirmedWithLeadManager),
    filled(section.sellingShareholderConsentsObtained),
    filled(section.anyExpectedPreIssueTransfers),
  ];
  const answered = core.filter(Boolean).length;

  const overlaysComplete = section.shareholderOverlays.every(
    (overlay) =>
      filled(overlay.shareholderId) &&
      (isFilledDecimal(overlay.sharesOfferedForSale) ||
        isFilledDecimal(overlay.otherExpectedPreIssueTransfer)),
  );
  const sellersComplete =
    section.sellingShareholderConsentsObtained !== 'yes' ||
    filled(section.sellingShareholderEligibilityConfirmed);
  const transfersExplained =
    section.anyExpectedPreIssueTransfers !== 'yes' ||
    filled(section.expectedPreIssueTransferDetails);

  return statusFrom(
    answered,
    core.length,
    overlaysComplete && sellersComplete && transfersExplained,
  );
}

export function evaluatePromoterContributionStatus(
  payload: CapitalOwnershipPayload,
): SectionStatus {
  const section = payload.promoterContributionLockInAndEncumbrances;
  const applicable = section.minimumPromoterContributionApplicable !== 'no';
  const core = [
    filled(section.minimumPromoterContributionApplicable),
    applicable
      ? section.contributionLots.length > 0
      : filled(section.exemptionFromMinimumContributionClaimed),
    filled(section.contributionBroughtInBeforeIssueOpening),
    filled(section.anyEncumbranceOnPromoterShares),
    filled(section.entirePreIssueCapitalLockInUnderstood),
    filled(section.sharesIneligibleForContributionExist),
  ];
  const answered = core.filter(Boolean).length;

  const lotsComplete = section.contributionLots.every(
    (lot) =>
      isFilledDecimal(lot.numberOfShares) &&
      filled(lot.dateOfAcquisition) &&
      filled(lot.modeOfAcquisition) &&
      filled(lot.eligibleForMinimumPromoterContribution),
  );
  const encumbrancesComplete = section.encumbrances.every(
    (item) => filled(item.encumbranceType) && isFilledDecimal(item.numberOfSharesEncumbered),
  );
  const encumbranceAnswered =
    section.anyEncumbranceOnPromoterShares !== 'yes' || section.encumbrances.length > 0;
  const ineligibleExplained =
    section.sharesIneligibleForContributionExist !== 'yes' ||
    filled(section.ineligibleSharesDetails);
  const exemptionExplained =
    section.exemptionFromMinimumContributionClaimed !== 'yes' || filled(section.exemptionBasis);

  return statusFrom(
    answered,
    core.length,
    lotsComplete &&
      encumbrancesComplete &&
      encumbranceAnswered &&
      ineligibleExplained &&
      exemptionExplained,
  );
}

export function evaluateOutstandingSecuritiesStatus(
  payload: CapitalOwnershipPayload,
): SectionStatus {
  const section = payload.outstandingSecuritiesTransactionsAndConfirmations;
  const confirmations = Object.values(section.confirmations);
  const confirmationsChecked = confirmations.filter(Boolean).length;

  const core = [
    filled(section.anyOutstandingConvertibleInstruments),
    filled(section.anyTransactionsInLastEighteenMonths),
    filled(section.allSharesDematerialisedBeforeFiling),
    filled(section.anyPendingShareTransfers),
    filled(section.anyDisputesOverTitleToShares),
    confirmationsChecked > 0,
  ];
  const answered = core.filter(Boolean).length;

  const instrumentsComplete =
    section.anyOutstandingConvertibleInstruments !== 'yes' ||
    (section.outstandingInstruments.length > 0 &&
      section.outstandingInstruments.every(
        (item) =>
          filled(item.instrumentType) &&
          isFilledDecimal(item.potentialEquitySharesOnConversion),
      ));
  const transactionsComplete =
    section.anyTransactionsInLastEighteenMonths !== 'yes' ||
    (section.recentTransactions.length > 0 &&
      section.recentTransactions.every(
        (item) =>
          filled(item.transactionDate) &&
          filled(item.transactionType) &&
          isFilledDecimal(item.numberOfShares),
      ));
  const pendingExplained =
    section.anyPendingShareTransfers !== 'yes' || filled(section.pendingShareTransferDetails);
  const disputesExplained =
    section.anyDisputesOverTitleToShares !== 'yes' || filled(section.titleDisputeDetails);
  const confirmationsComplete = confirmationsChecked === confirmations.length;

  return statusFrom(
    answered,
    core.length,
    instrumentsComplete &&
      transactionsComplete &&
      pendingExplained &&
      disputesExplained &&
      confirmationsComplete,
  );
}

export function calculateCapitalOwnershipProgress(
  payload: CapitalOwnershipPayload,
): CapitalOwnershipProgress {
  const sections: Record<CapitalOwnershipSectionId, SectionStatus> = {
    'current-capital-structure': evaluateCurrentCapitalStructureStatus(payload),
    'share-capital-history': evaluateShareCapitalHistoryStatus(payload),
    'shareholders-beneficial-ownership': evaluateShareholdersStatus(payload),
    'promoters-and-control': evaluatePromotersAndControlStatus(payload),
    'pre-post-issue-ownership': evaluatePrePostIssueStatus(payload),
    'promoter-contribution-lock-in': evaluatePromoterContributionStatus(payload),
    'outstanding-securities-confirmations': evaluateOutstandingSecuritiesStatus(payload),
  };

  const statuses = Object.values(sections);
  const sectionsComplete = statuses.filter((status) => status === 'complete').length;
  const totalSections = statuses.length;
  let overallStatus: SectionStatus = 'not_started';
  if (sectionsComplete === totalSections) overallStatus = 'complete';
  else if (statuses.some((status) => status !== 'not_started')) overallStatus = 'in_progress';

  return { sections, sectionsComplete, totalSections, overallStatus };
}

export function listIncompleteCapitalOwnershipSections(
  payload: CapitalOwnershipPayload,
): string[] {
  const progress = calculateCapitalOwnershipProgress(payload);
  const incomplete: string[] = [];
  for (const [id, status] of Object.entries(progress.sections) as Array<
    [CapitalOwnershipSectionId, SectionStatus]
  >) {
    if (status !== 'complete') {
      incomplete.push(`${CAPITAL_OWNERSHIP_SECTION_LABELS[id]} incomplete`);
    }
  }
  return incomplete;
}
